sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,635 @@
+
+    @Nullable abstract ConnectionConfiguration getConnectionConfiguration();
+    abstract long getBatchSize();
+    abstract int getBatchSizeMegaBytes();
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration connectionConfiguration);
+      abstract Builder setBatchSize(long batchSize);
+      abstract Builder setBatchSizeMegaBytes(int batchSizeMegaBytes);
+      abstract Write build();
+    }
+
+    public Write withConnectionConfiguration(ConnectionConfiguration connectionConfiguration) {
+      checkArgument(connectionConfiguration != null, "ElasticsearchIO.write()"
+          + ".withConnectionConfiguration(configuration) called with null configuration");
+      return builder().setConnectionConfiguration(connectionConfiguration).build();
+    }
+
+    public Write withBatchSize(long batchSize) {
+      return builder().setBatchSize(batchSize).build();
+    }
+
+    public Write withBatchSizeMegaBytes(int batchSizeMegaBytes) {
+      return builder().setBatchSizeMegaBytes(batchSizeMegaBytes).build();
+    }
+
+    @Override
+    public void validate(PCollection<String> input) {
+      checkState(getConnectionConfiguration() != null, "ElasticsearchIO.write() requires a "
+          + "connection configuration to be set via withConnectionConfiguration(configuration)");
+    }
+
+    @Override
+    public PDone apply(PCollection<String> input) {
+      input.apply(ParDo.of(new WriterFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    private static class WriterFn extends DoFn<String, Void> {
+
+      private final Write spec;
+
+      private transient RestClient restClient;
+      private ArrayList<String> batch;
+      private long currentBatchSizeBytes;
+
+      public WriterFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void createClient() throws Exception {
+        restClient = spec.getConnectionConfiguration().createClient();
+      }
+
+      @StartBundle
+      public void startBundle(Context context) throws Exception {
+        batch = new ArrayList<>();
+        currentBatchSizeBytes = 0;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        String document = context.element();
+        batch.add(String.format("{ \"index\" : {} }%n%s%n", document));
+        currentBatchSizeBytes += document.getBytes().length;
+        if (batch.size() >= spec.getBatchSize()
+            || currentBatchSizeBytes >= (spec.getBatchSizeMegaBytes() * 1024L * 1024L)) {
+          finishBundle(context);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle(Context context) throws Exception {
+        StringBuilder bulkRequest = new StringBuilder();
+        if (batch.size() > 0) {
+          for (String json : batch) {
+            bulkRequest.append(json);
```

Comment:
Actually, if a batch contains exactly batchSize documents or batchSizeMagabytes bytes, finishBundle() gets called one more time after that batch with an empty batch.  insert with empty batch in ES do not work. I protected with if() to avoid bulk request when batch is empty. 

For the sizes removal: well, I agree batchSize in number of docs is not so relevant for performance (because doc size changes), but batchSizeInMegabytes is. As explained in the above doc, end user might need to configure it to get best performance. And I remember that, as a user, I had to do this in a project to improve ES indexation time. The fact is that both are present is ES bulk API. [https://www.elastic.co/guide/en/elasticsearch/client/java-api/2.4/java-docs-bulk-processor.html](url).See setBulkActions and setBulkSize.

Besides, in the IO I set defaults to 1000 docs and 5 MB